### PR TITLE
[SOAR-14571] - Zoom - Delete User Returns Json Parsing Error

### PR DESCRIPTION
**Pull request by TKuzas-R7**
### Description

Describe the proposed changes:

  - Moved check for 204 status code above checks for 4xx status codes to avoid json parsing errors.
  - removed mapping of status codes so that .json() is only called once
  - Added Examples to plugin spec
  - fixed failing unit test
  - updated imports for unit tests

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]
Looks great, have a few comments. Can I also check that we've a test case for the new error code in our unit tests?

File: plugins/zoom/Dockerfile [hunk 1]
```diff
@@ -1,24 +1,18 @@
 FROM rapid7/insightconnect-python-3-38-plugin:4
```
Comment (igorski-r7) [inline plugins/zoom/Dockerfile:1]: That's interesting. Does it really have to be the version `4` of the SDK? Or should it be `5` 🤔 
@mrinehart-r7

File: plugins/zoom/help.md [hunk 2]
```diff
@@ -20,51 +20,52 @@ Business, or Enterprise plan.
...
-
+  
 Example input:
 
 ```
 {
   "account_id": "dBs0x4Kf7HuIK0LLbzMduW",
   "authentication_retry_limit": 5,
-  "client_id": "9de5069c5afe602b2ea0a04b66beb2c0",
+  "client_id": "{"secretKey": "9de5069c5afe602b2ea0a04b66beb2c0"}",
```
Comment (igorski-r7) [inline plugins/zoom/help.md:45]: Please parse it as `dict` instead of `str` 😄

File: plugins/zoom/help.md [hunk 3]
```diff
@@ -20,51 +20,52 @@ Business, or Enterprise plan.
...
 ## Technical Details
 
 ### Actions
+  
 
 #### Create User
-
-This action is used to create user associated to account.
+  
+Create user associated to account
```
Comment (igorski-r7) [inline plugins/zoom/help.md:57]: Wondering about prefix. That looked better with it
```suggestion
This action is used to create user associated to account.  
```

File: plugins/zoom/help.md [hunk 4]
```diff
@@ -125,125 +126,103 @@ Example input:
...
-    "verified": 0,
-    "created_at": "2020-06-03T21:30:52Z",
-    "first_name": "",
-    "language": "",
-    "phone_number": "",
-    "group_ids": [],
-    "last_name": "",
-    "status": "pending"
-  }
+  "user": {}
```
Comment (igorski-r7) [inline plugins/zoom/help.md:173]: We lost a couple of examples here. They are all need to be added to `plugin.spec.yaml` file

**Review by igorski-r7 — CHANGES_REQUESTED** [review]

File: plugins/zoom/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (TKuzas-R7) [reply plugins/zoom/Dockerfile:1]: Tasks Fail with version 5.

File: plugins/zoom/plugin.spec.yaml [hunk 5]
```diff
@@ -55,55 +59,65 @@ types:
...
       description: The type of client of the user's device
       type: string
       required: false
+      example: Browser
     version:
       title: Version
       description: The version of the client of the user's device
       type: string
       required: false
+      example: "-"
```
[the remark this answers is not in the document]
Comment (TKuzas-R7) [reply plugins/zoom/plugin.spec.yaml:95]: When testing, "-" is the output i get.

File: plugins/zoom/icon_zoom/util/api.py [hunk 6]
```diff
@@ -193,27 +193,25 @@ def _handle_response(self, response: Response, allow_404: bool, original_call_ar
...
-                if status_code == 404 and allow_404:
-                    return None
-                raise exception
+        if response.status_code == 400:
+            raise PluginException(preset=PluginException.Preset.BAD_REQUEST, data=response.json())
+        if response.status_code == 404:
+            if allow_404:
+                return None
+            raise PluginException(preset=PluginException.Preset.NOT_FOUND, data=response.json())
+        if response.status_code >= 409:
```
[the remark this answers is not in the document]
Comment (TKuzas-R7) [reply plugins/zoom/icon_zoom/util/api.py:209]: good catch. I'm not quite sure why i included the greater than

**Comment by TKuzas-R7** [comment]
> Looks great, have a few comments. Can I also check that we've a test case for the new error code in our unit tests?

We dont. but there will be an integration test to cover it.

**Comment by TKuzas-R7** [comment]
> 

There is a test for it yes,
the unit test didnt catch the issue because it was being mocked.
It looks like the Integration tests do a better job finding the issues.

File: plugins/zoom/Dockerfile (hunk 1 above)
Comment (TKuzas-R7) [reply plugins/zoom/Dockerfile:1]: attempting to run the task in postman after updating SDK to version 5 returns the following error:
`{
        "error": "not enough values to unpack (expected 3, got 2)",
        "log": "rapid7/Zoom:4.0.2. Step name: monitor_sign_in_out_activity\nFirst run\nGetting events from 2023-06-26T15:15:28Z until 2023-06-27T15:15:28Z\nCalling GET https://api.zoom.us/v2/report/activities\nGot response status code: 401\nReceived HTTP 401, re-authenticating to Zoom...\nCalling Zoom API to refresh OAuth token...\nGot status code 200 from OAuth token refresh\nRequest for new OAuth token was successful!\nCalling GET https://api.zoom.us/v2/report/activities\nGot response status code: 200\nGot 4 events from Zoom!\nLatest event time is: 2023-06-27T14:26:52Z\nUpdated state, state is now: {'boundary_events': ['9f683d4d635a549ddf152d0b810bbd82a9ce33ff'], 'last_request_timestamp': '2023-06-27T15:15:28Z', 'latest_event_timestamp': '2023-06-27T14:26:52Z', 'status_code': 200}\nnot enough values to unpack (expected 3, got 2)\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-5.1.4-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 424, in handle_step\n    output, state, has_more_pages = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-5.1.4-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 582, in start_step\n    output, state, has_more_pages = func(params, state)\nValueError: not enough values to unpack (expected 3, got 2)\n",
        "meta": {},
        "status": "error"
    }`

File: plugins/zoom/plugin.spec.yaml [hunk 7]
```diff
@@ -146,87 +166,100 @@ types:
...
       title: Status
       description: Status of user
       type: string
       required: false
+      example: active
 actions:
   get_user:
     title: Get User
-    description: Get user in Zoom account
+    description: This action is used to get user in Zoom account
```
Comment (cmcnally-r7) [inline plugins/zoom/plugin.spec.yaml:245]: hey so for now dont worry about adding the prefix here. Keep the description the same, we're gonna get the tooling to handle adding the prefix just for the help.md. Best to keep these descriptions succinct like what it was before this change

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]
one small thing

File: plugins/zoom/Dockerfile (hunk 1 above)
Comment (mrinehart-r7) [reply plugins/zoom/Dockerfile:1]: Do not update this to version 4 yet - the code isn't ready to support the `has_more_pages` return output yet.

File: plugins/zoom/plugin.spec.yaml [hunk 8]
```diff
@@ -20,26 +20,30 @@ resources:
   license_url: https://github.com/rapid7/insightconnect-plugins/blob/master/LICENSE
   vendor_url: https://www.zoom.com/
 enable_cache: false
+sdk:
+  type: full
+  version: 4
```
[the remark this answers is not in the document]
Comment (mrinehart-r7) [reply plugins/zoom/plugin.spec.yaml:25]: This can't be updated to version 5 until has_more_pages output is supported for the task.

**Review by ablakley-r7 — APPROVED** [review]

**Review by cmcnally-r7 — APPROVED** [review]

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**